fortune-mod/util/randstr.man
.\"
.\" Randstr: grab a random text string from a specified text file
.\" Amy A. Lewis, October, 1995
.\"
.TH RANDSTR 1 "October 1995" "Linux hackery"
.SH NAME
randstr \- fetch a random text string from a specified file
.SH SYNOPIS
.BI "randstr " filename
.SH DESCRIPTION
In essence,
.B randstr
is a stripped-down, minimalist version of the popular
.BR fortune (6)
program.
.PP
It reads files with the same structure as the fortunes databases, and
displays a random string.
.SS Usage
.B randstr
might be used, with a database of signatures created with a text editor
and the
.BR strfile "(8) and " unstr (8)
utilities, to randomly change the
.I $HOME/.signature
file.  To do so, create the necessary database as, for example,
.IR signatures " and " signatures.dat
in the home directory, and add the following line to
.IR .profile " or " .login :
.RS
.BI "randstr " signatures
.RI > .signature
.RE
.PP
As superuser, a similar sort of thing could be placed in the
.BR Pnews "(1) or " Rnmail (1)
scripts, although it should certainly test to be certain that the files
exist, and creating the files of signatures might require a good deal of
help for a lot of users.
.PP
Also as superuser, one might use
.B randstr
with a shell or
.B Perl
script to read the
.I etc/passwd
database and create a
.I strfile
type database, and use this database to run a monthly lottery.
.SH BUGS
None known.
.SH SEE ALSO
.BR fortune "(6), " strfile (1)


fortune-mod/ChangeLog
(Note: this file has been re-arranged to be in reverse chronological
order, which is The Right Thing for ChangeLogs - DLC)

March 05, 2004 (fortune-mod-1.99.1)

	Most of the changes have occured at some point in time in the last
	5 years.

	A high number of spelling, punctuation, formatting and grammar
	fixes.

	Internationalisation support.

	New -c option to see which file a fortune came from.

August 29, 1997

  Incorporated a couple of minor changes made in the (old) Debian
  fortune-mod package, including the addition of an extra data file
  called 'cookie'.  Renamed some documentation files, and included some
  install information for non-Linux users.  Added a "-v" option to
  report the program version.

  I intend to submit this distribution to SunSITE RSN.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

May, 1997

  This release fixes many of the portability problems with the
  fortune-mod program released by Amy Lewis in October, 1995.  The
  previous version had many Linux-isms in it, which left it unworkable
  on any other platform.  This version replaces most of these with more
  standard calls, making it more likely to work under other platforms.
  The Makefiles have been modified so that GNU's gcc and make are no
  longer required: any standard make and ANSI-compatible C compiler
  should work.  Sorry, pre-ANSI compilers are not supported (c'mon, this
  is the 90's, darn it!)

  This version has been tested to work on SunOS 4.1.x as well as Linux.
  All changes made to the sources were as platform-independant as
  possible.  Therefore, no "#ifdef LINUX" or "#ifdef SUNOS4" directives
  appear in the code.  An effect of this is that a number of 'implicit
  declaration' warnings are emitted by gcc under SunOS 4.1.x, but this a
  problem with SunOS's standard headers, not with the program or the
  compiler.  The benefit of this approach is that it eases the work of
  expanding the port to include other platforms. Bug reports and fixes
  for other platforms are most welcome!

  A few "standard" C function calls were replaced with more standard
  counterparts at various points.  Generally, when there was a choice
  between a BSD version of a function and a POSIX version, the POSIX
  version was favoured (even though fortune originated on BSD).  An
  exception to this was the regex functions: either POSIX or BSD
  versions can be used, with selection made via the top-level Makefile.

  Fortune and strfile also compiles on Solaris 2.5, but a discrepency
  between the declaration and implementation of 'struct dirent' on the
  test platform caused fortune to execute incorrectly there.  It is not
  certain whether this is a bug in the header file, the C library, or
  the test platform.

  While the making of this release was not meant to become a bug
  search-and-destroy mission, some bugs were inadvertantly discovered
  and fixed. including the known bug of using -a with a file that occurs
  in both the offensive and inoffensive directories.  Fixing this
  particular bug required a way to be able to seperately identify two
  fortune files with the same name, with one in the inoffensive
  directory, the other one in the offensive directory.  Now, such a name
  will be taken to be the inoffensive file by default.  However, you can
  now append '-o' to a fortune name, and the '-o' will be removed and
  the offensive directory will be searched.  Thus you can say (assuming
  you use the distributed datfiles):

        fortune 80% politics politics-o

  Which has an 80% change of giving an "inoffensive" political fortune,
  and a 20% change of giving an "offensive" political one.  Note that
  this makes fortune-mod backwards-compatible with BSD fortune, but only
  for users, not fortune database maintainers.

  Of course, this solution only passes the buck: with the above example
  again, if you have a 'politics-o' file in your inoffensive directory,
  you are back to square one.  OTOH, seeing that '-o' originally was
  meant for offensive fortunes, using it for inoffensive ones is simply
  asking for trouble.

  Bug fixed: Fortune's definition of a fortune length (for -s and -l)
  was inconsistant.  Unordered fortunes counted the 2 delimiting
  characters (as hinted in the man page), but sorted or randomized
  fortunes did not.  Now the delimiting characters are _NEVER_ counted,
  so you will always get the length limit you expect.

  Another bug fix: -l and -s can now work together with -m.  Previously
  -l and -s were ignored when -m was in effect.  The new behaviour helps
  me count how many long or short fortunes there are in a file.

  Ansify has been removed from the package, as well as some filter
  scripts from NetBSD that no longer appear useful.  Randstr has been
  kept, but has not been improved in any way.  It at least has a man
  page: maybe somebody will find it useful.

  Some of the documentation (including the man pages) has been improved
  and updated, and some files have been renamed so that the package
  looks less Linux-specific.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

Late October 1995
  Ansify has been abandoned.  I'm going to distribute this working version
  of fortune, and then see how difficult it would be to add termcap/terminfo
  enhancements to fortune itself--I don't anticipate serious problems, but
  I'd rather go ahead and get this on the net.

  A last-minute change was made to the way that percentages are displayed
  with -f; it is now in the format nnn.nn%.  The reason for this is that
  with the multiplication of small files, fortune -af displayed a large
  number of "0%"s--no worse than the old version, but not helpful.

  The fortunes database was finally cleaned up, and this version is now
  being distributed (at least, I hope it is).  I don't consider the current
  division of fortunes among files absolutely canonical; some are certainly
  in the wrong places.  But things are *better*.

Amy A. Lewis	alewis@email.unc.edu


fortune-mod/util/randstr.c
/*
 * The following code has been derived chiefly from the BSD distributions
 * of the utility program unstr and the random quote displayer fortune.
 * The utility produced by this code shares characteristics of both
 * (it might be regarded as a minimalist implementation of fortune, for
 * large values of 'minimal'), and is here offered so as to have the
 * minimal necessary tools for rabbiting a strfile routine into some other,
 * more significant program, all in one place.  A programmer who cares and
 * has the proper training could probably clean this up significantly; it's
 * all stolen code (first rule of programming: steal) hacked together to
 * fit.  Or, to paraphrase the old saw about how the British built ships,
 * it's coded by the mile and cut off to order.  In that analogy, this
 * program's about an inch--and separated with an axe.
 *
 * Axe murderess programming.  Wotta concept!
 *
 * Use at your own peril, especially as a pattern (kludge, kludge!). This
 * program, at least, shouldn't have any real chance of corrupting data,
 * though; it opens files ro and dumps to the screen.  If you redirect
 * output, you definitely do so at your own peril (I lost six hours of
 * editing on a fortune file that way, by redirecting the output of unstr
 * before it had an outputfile option, trying to skip over the mv x.sorted
 * x step.  Axe murderess redirection, in that case).
 *
 * Blame Amy A. Lewis.  September, 1995.  alewis@email.unc.edu
 */

/*-
 * Copyright (c) 1991, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ken Arnold.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* randstr repeats the minimum functionality of the fortune program.  It
 * finds the fortune text or data file specified on the command line --
 * one or the other, not both -- generates a random number, and displays
 * the text string indexed.  No provision is made for any other command
 * line switches.  At all.
 *
 * Usage:
 *
 * randstr filename[.ext]
 *
 * Example: run sed or Perl over your /etc/passwd, and kick out a
 * strfile-format file containing lognames on the first line and full
 * names on the second.  Write a script called 'lottery' which is
 * called once a month from crontab; it in turn calls randstr lusers,
 * and the winning luser gets a prize notification sent by email from
 * the lottery script.  Living up to promises is optional.
 *
 * Note: if you're a sysadmin who regularly reads _Mein Kampf_ for the
 * deep truths buried in it, and believe in Truth, Justice, and the
 * American Family, you could use this to replace fortune, by pointing
 * it at a small, Family Values database.  The great advantage to this,
 * in my opinion, is that it wouldn't take up any disk space at all.
 * Who're you gonna quote?  Dan Quayle?
 */

#include        <netinet/in.h>
#include        <sys/param.h>
#include        "strfile.h"
#include        <stdio.h>
#include        <stdlib.h>
#include        <ctype.h>
#include        <string.h>
#include        <unistd.h>
#include        <time.h>
#ifndef MAXPATHLEN
#define MAXPATHLEN      1024
#endif /* MAXPATHLEN */

char *Infile,                   /* name of input file */
  Datafile[MAXPATHLEN],         /* name of data file */
  Delimch;                      /* delimiter character */

FILE *Inf, *Dataf, *Outf;

off_t pos, Seekpts[2];          /* seek pointers to fortunes */


void getargs(int ac, char *av[])
{
    extern int optind;
    char *extc;

    av += optind + 1;

    if (*av)
    {
        Infile = *av;
/* Hmm.  Don't output anything if we can help it.
 * fprintf(stderr, "Input file: %s\n",Infile); */
        if (!strrchr(Infile, '.'))
        {
            strcpy(Datafile, Infile);
            strcat(Datafile, ".dat");
        }
        else
        {
            strcpy(Datafile, Infile);
            extc = strrchr(Infile, '.');
            *extc = '\0';
        }
    }
    else
/*    {
 * Don't write out errors here, either; trust in exit codes and sh
 * fprintf(stderr, "No input file name\n");
 * fprintf(stderr, "Usage:\n\tunstr [-c C] datafile[.ext] [outputfile]\n");
 */ exit(1);
/*    } */
}

/*
 * get_pos:
 *      Get the position from the pos file, if there is one.  If not,
 *      return a random number.
 */
void get_pos(STRFILE * fp)
{
    pos = random() % fp->str_numstr;
    if (++(pos) >= fp->str_numstr)
        pos -= fp->str_numstr;
}

/*
 * get_fort:
 *      Get the fortune data file's seek pointer for the next fortune.
 */
void get_fort(STRFILE fp)
{
    register int choice;

    choice = random() % fp.str_numstr;

    get_pos(&fp);
    fseek(Dataf, (long) (sizeof fp + pos * sizeof Seekpts[0]), 0);
    fread(Seekpts, sizeof Seekpts, 1, Dataf);
    Seekpts[0] = ntohl(Seekpts[0]);
    Seekpts[1] = ntohl(Seekpts[1]);
}

void display(FILE * fp, STRFILE table)
{
    register char *p, ch;
    unsigned char line[BUFSIZ];
    int i;

    fseek(fp, (long) Seekpts[0], 0);
    for (i = 0; fgets(line, sizeof line, fp) != NULL &&
         !STR_ENDSTRING(line, table); i++)
    {
        if (table.str_flags & STR_ROTATED)
            for (p = line; (ch = *p); ++p)
                if (isupper(ch))
                    *p = 'A' + (ch - 'A' + 13) % 26;
                else if (islower(ch))
                    *p = 'a' + (ch - 'a' + 13) % 26;
        fputs(line, stdout);
    }
    fflush(stdout);
}

int main(int ac, char **av)
{
    static STRFILE tbl;         /* description table */

    getargs(ac, av);
    if ((Inf = fopen(Infile, "r")) == NULL)
    {
        perror(Infile);
        exit(1);
    }
    if ((Dataf = fopen(Datafile, "r")) == NULL)
    {
        perror(Datafile);
        exit(1);
    }
    fread((char *) &tbl, sizeof tbl, 1, Dataf);
    tbl.str_version = ntohl(tbl.str_version);
    tbl.str_numstr = ntohl(tbl.str_numstr);
    tbl.str_longlen = ntohl(tbl.str_longlen);
    tbl.str_shortlen = ntohl(tbl.str_shortlen);
    tbl.str_flags = ntohl(tbl.str_flags);

    srandom((int) (time((time_t *) NULL) + getpid()));
    get_fort(tbl);
    display(Inf, tbl);

    exit(0);

    fclose(Inf);
    fclose(Dataf);
    fclose(Outf);
    exit(0);
}
